Neos.Media.Browser/Classes/Controller/AssetController.php

```diff
@@ -854,9 +799,158 @@ public function addFlashMessage($messageBody, $messageTitle = '', $severity = Me
     }
+
+    /**
+     * @param string $view
+     * @param string $sortBy
+     * @param string $sortDirection
+     * @param string $filter
+     */
+    private function applyViewOptionsToBrowserState(string $view = null, string $sortBy = null, string $sortDirection = null, string $filter = null): void
+    {
+        if (!empty($view)) {
+            $this->browserState->set('view', $view);
+        }
+        if (!empty($sortBy)) {
+            $this->browserState->set('sortBy', $sortBy);
+        }
+        if (!empty($sortDirection)) {
+            $this->browserState->set('sortDirection', $sortDirection);
+        }
+        if (!empty($filter)) {
+            $this->browserState->set('filter', $filter);
+        }
+
+        foreach (['view', 'sortBy', 'sortDirection', 'filter'] as $optionName) {
+            $this->view->assign($optionName, $this->browserState->get($optionName));
+        }
+    }
+
+    /**
+     * @param $assetSourceIdentifier
+     */
+    private function applyActiveAssetSourceToBrowserState($assetSourceIdentifier)
+    {
+        if ($assetSourceIdentifier !== null && isset($this->assetSources[$assetSourceIdentifier])) {
+            $this->browserState->setActiveAssetSourceIdentifier($assetSourceIdentifier);
+        }
+    }
+
+    /**
+     * @param int $tagMode
+     * @param Tag $tag
+     * @param AssetCollection|null $activeAssetCollection
+     */
+    private function applyTagToBrowserState(int $tagMode = null, Tag $tag = null, AssetCollection $activeAssetCollection = null): void
+    {
+        if ($tagMode === self::TAG_GIVEN && $tag !== null) {
+            $this->browserState->set('activeTag', $tag);
+            $this->view->assign('activeTag', $tag);
+        } elseif ($tagMode === self::TAG_NONE || $tagMode === self::TAG_ALL) {
+            $this->browserState->set('activeTag', null);
+            $this->view->assign('activeTag', null);
+        }
+        $this->browserState->set('tagMode', $tagMode);
+
+        // Unset active tag if it isn't available in the active asset collection
+        if ($this->browserState->get('activeTag') && $activeAssetCollection !== null && !$activeAssetCollection->getTags()->contains($this->browserState->get('activeTag'))) {
+            $this->browserState->set('activeTag', null);
+            $this->view->assign('activeTag', null);
+        }
+
+        if (!$this->browserState->get('activeTag') && $this->browserState->get('tagMode') === self::TAG_GIVEN) {
+            $this->browserState->set('tagMode', self::TAG_ALL);
+        }
+    }
+
+    /**
+     * @return AssetSource
+     */
+    private function getAssetSourceFromBrowserState(): AssetSource
+    {
+        $assetSourceIdentifier = $this->browserState->getActiveAssetSourceIdentifier();
+        if (!isset($this->assetSources[$assetSourceIdentifier])) {
+            $assetSourceIdentifiers = array_keys($this->assetSources);
+            $assetSourceIdentifier = reset($assetSourceIdentifiers);
```

Comment:
Don't we want to error here if there was an `$assetSourceIdentifier` set that was not found?

Reply:
I'd say, rather not show an error. Because it could be that asset source have been reconfigured and you still have an old identifier in your session - then falling back to another asset source would be okay, I guess.